### PR TITLE
Optimization and rationalization of horizontal spacing calculations

### PR DESCRIPTION
**Pull request by mike-spa**
Mostly, a new algorithm for system justification, which was the most expensive function of the whole layout. This allows to save up to 60% of the total layout time. The program now feels noticeably more responsive. Besides that, some general cleanup of the code.
All vtests are expected to fail because of rounding errors, but otherwise this system is meant to produce identical results.

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

File: src/engraving/libmscore/segment.h [hunk 1]
```diff
@@ -78,6 +88,7 @@ class Segment final : public EngravingItem
     Fraction _ticks;   // { Fraction(0, 1) };
     Spatium _extraLeadingSpace;
     double _stretch;
+    double _widthOffset = 0; // part of the segment width that will not be stretched during system justification
```
[the remark this answers is not in the document]
Comment (vpereverzev) [reply src/engraving/libmscore/segment.h:91]: I suggest to go through your changes and make sure tha you're initializing "double/float" values properly

File: src/plugins/api/elements.h [hunk 2]
```diff
@@ -189,7 +189,6 @@ class EngravingItem : public mu::engraving::PluginAPI::ScoreElement
     API_PROPERTY(noStem,                  NO_STEM)
     API_PROPERTY(slurDirection,           SLUR_DIRECTION)
     API_PROPERTY(leadingSpace,            LEADING_SPACE)
-    API_PROPERTY(distribute,              DISTRIBUTE)
```
Comment (vpereverzev) [inline src/plugins/api/elements.h:192]: why did this dissapear?

File: src/engraving/libmscore/segment.cpp [hunk 3]
```diff
@@ -2782,4 +2782,104 @@ void Segment::computeCrossBeamType(Segment* nextSeg)
...
+/***********************************************
+ * stretchSegmentsToWidth
+ * Stretch a group of (chordRestType) segments
+ * by the specified amount. Uses the spring-rod
+ * method.
+ * *********************************************/
+
+void Segment::stretchSegmentsToWidth(std::vector<Spring>& springs, double width)
+{
+    if (width <= 0) {
```
[the remark this answers is not in the document]
Comment (vpereverzev) [reply src/engraving/libmscore/segment.cpp:2795]: also, go through your changes and make sure that your comparisons of double/float values are properly written

File: src/engraving/libmscore/segment.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mike-spa) [reply src/engraving/libmscore/segment.h]: Done! Hope I din't forget any

File: src/engraving/libmscore/segment.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mike-spa) [reply src/engraving/libmscore/segment.cpp]: Done!

File: src/engraving/layout/layoutsystem.cpp [hunk 4]
```diff
@@ -376,45 +379,58 @@ System* LayoutSystem::collectSystem(const LayoutOptions& options, LayoutContext&
...
+        }
+        Measure* m = toMeasure(mb);
+        double oldWidth = m->width();
+        // If system is currently larger than margin (because of acceptanceRange) compute width
+        // with a reduced pre-stretch, because justifySystem expects curSysWidth < targetWidth
+        if (targetSystemWidth > curSysWidth) {
+            m->computeWidth(minTicks, maxTicks, preStretch);
+        } else {
+            m->computeWidth(minTicks, maxTicks, preStretch);
+        }
```
Comment (vpereverzev) [inline src/engraving/layout/layoutsystem.cpp:425]: logic in both "if" and "else" seems to be equal

**Review by vpereverzev — APPROVED** [review]

**Pull request merged**